Improve exception logging for SMTP Notify

ThioJoe opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
For the smtp.notify component, adds handling for each type of smtplib exception found [here](https://docs.python.org/3/library/smtplib.html), and creates a log warning and/or debug message for each, which contains details such as returned error code and message. 

This will help users to diagnose issues with their SMTP configuration. I was having the same problem as the user [here](https://community.home-assistant.io/t/how-to-debug-notification-error-smtpexception-sending-email/383368), and I imagine many others have too. The current handling gives no information about the nature of the error if there is one, and does not give the returned error message. But this pull request will provide that information.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

ThioJoe committed:
Update Notify.py - Add logging for SMTP Exceptions

Add handling for each type of smtplib exception, and log a warning and/or debug message for each, and containing details such as error code and message. This will help users to diagnose issues with their SMTP configuration.

MarkGodwin commented on homeassistant/components/smtp/notify.py:
```diff
@@ -223,14 +223,80 @@ def _send_email(self, msg, recipients):
+                _LOGGER.debug("SMTPRecipientsRefused Full Error Contents: %s", err)
+                mail.quit()
+                mail = self.connect()
+            except smtplib.SMTPDataError as err:
```
A lot of these catch clauses are very similar. You could catch many of them in a single except block (or perhaps just catch `smtplib.SMTPResponseException` which all the exceptions with `smtp_code` and `smtp_error` derive from), and include the exception type name in the formatted strings (e.g. `type(err).__name__`) without losing too much logging information, and reducing the verbosity of the code a lot.

ThioJoe replied:
You're right, I was able to drastically reduce the code by just catching them all at once and dynamically setting the messages based on the exceptions. I figured I'd still catch the four that have `smtplib.SMTPResponseException` as their base exception so that I could reference them specifically by name.

gjohansson-ST reviewed (changes requested):
Looks good but as already mentioned try to consolidate errors that provide same logging.

ThioJoe committed:
Consolidate smtplib Exception Clauses

Instead of catching the smtplib exceptions individually with separate except clauses, catch them all at once and output messages dynamically based on the type and contents of the exception.

ThioJoe commented:
> Looks good but as already mentioned try to consolidate errors that provide same logging.

Yep I was able to reduce the code drastically by consolidating

raman325 commented on homeassistant/components/smtp/notify.py:
```diff
@@ -223,14 +223,31 @@ def _send_email(self, msg, recipients):
-                _LOGGER.warning(
-                    "SMTPServerDisconnected sending mail: retrying connection"
+            except (
+                smtplib.SMTPSenderRefused,
+                smtplib.SMTPDataError,
+                smtplib.SMTPConnectError,
+                smtplib.SMTPHeloError,
+                smtplib.SMTPResponseException,
+                smtplib.SMTPServerDisconnected,
+                smtplib.SMTPRecipientsRefused,
+                smtplib.SMTPNotSupportedError,
+                smtplib.SMTPException,
```
`SMTPException` is the base exception class, so you don't need to enumerate all of these exceptions, you can just have the generic one, and `__name__` should still return the correct class name

ThioJoe replied:
You're right and I had indeed considered that, but I decided the extra few lines were worth the added readability and maintainability. 

Enumerating them all could make it easier for others in the future to understand which exceptions were considered when I wrote the code, so adding additional functionality for one of the specific exceptions might be easier, if for example they could instantly know the code does not handle some other exception.

Only catching the base SMTPException could make the code less clear. This was effectively the problem I was remedying in the first place, since the original code just showed it catching a single exception and it took extra time to figure out how it related to the other ones.

gjohansson-ST replied:
@raman325 is probably right here to catch only the base exception as if other errors are added down the line it's still going to get caught as the base exception is in the list.
Don't think the list of exceptions provides any necessary value as you want to capture all of them anyhow to provide to logger output.

raman325 replied:
agreed with @gjohansson-ST - it's clear from the code that you've made the exception logging generic for all types. If you are concerned about readability, you can also add a comment that just says this is the base exception class and that details will be logged, but it took me all of 30 seconds to understand that was your intention.

raman325 reviewed (changes requested):
marking this as draft per above conversation

raman325 reviewed (changes requested):
marking this as draft per above conversation
